### ISSUE TITLE
Review popular cache plugins and their dynamic caching opportunities

### ISSUE THREAD

**Issue by duracelltomi** [duracelltomi/gtm4wp#132]
Moving conversation from https://github.com/duracelltomi/gtm4wp/issues/100 into this new thread

A general issue with an active cache plugin is that some dynamic parts of the data layer will be cached as well: geo data, weather data, etc

As far as I know some cache plugins already have dynamic caching capabilities while others do not have this functionality yet.

**Comment by duracelltomi** [comment]
W3 Total Cache: https://github.com/W3EDGE/w3-total-cache/wiki/FAQ:-Developers#how-does-page-fragment-cache-work

**Comment by duracelltomi** [comment]
WP Super Cache: https://github.com/Automattic/wp-super-cache/blob/master/plugins/dynamic-cache-test.php

**Comment by duracelltomi** [comment]
I would be happy if you could share your findings with WP Rocket as currently I was not able to set it up so that only the dataLayer part of the page is not cached.

**Comment by svaldesm** [comment]
As far as I can tell (I just tested it), WP Rocket does not interfere with the dataLayer. 
I tried with the option to cache for logged-in users and without it. 
It wasn't a huge test but at least I could tell that the issues I'm having are coming from Varnish. I'm available if you want to do further testings.

**Comment by svaldesm** [comment]
As far as I can tell (I just tested it), WP Rocket does not interfere with the dataLayer. 
I tried with the option to cache for logged-in users and without it. 
It wasn't a huge test but at least I could tell that the issues I'm having are coming from Varnish. I'm available if you want to do further testings.

**Comment by duracelltomi** [comment]
Coming back to this — closing it as resolved by the work on #398 (the cache-safe data layer), which took a different, cache-agnostic route rather than integrating with each cache plugin's dynamic-caching API.

Instead of hooking every cache plugin's fragment/ESI mechanism, 2.0 makes the plugin cache-agnostic: the visitor/session-specific data layer values (visitor IP, Cloudflare country, logged-in user data, WooCommerce customer/cart, and the one-shot add-to-cart / purchase events) are no longer rendered into the cached HTML — they're delivered client-side (browser-known values need no request; the rest are fetched once per session or gated by a change/login/event cookie, never per page). So a page cached by LiteSpeed / WP Rocket / Varnish / Cloudflare no longer leaks one visitor's data to another, without GTM4WP needing to know which cache sits in front of it. It's opt-in via the experimental "Cache-safe data layer" option.

The other dynamic values this issue mentioned — geo and weather data — were removed in 2.0, so they're no longer a caching concern either.

Closing as superseded by #398; ships in 2.0 beta1.
